// guilded.js/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ROUTES = exports.constants = void 0;
exports.constants = {
    clientEvents: {
        CALENDAR_EVENT_CREATED: "calendarEventCreated",
        CALENDAR_EVENT_UPDATED: "calendarEventUpdated",
        CALENDAR_EVENT_DELETED: "calendarEventDeleted",
        CALENDAR_EVENT_RSVP_UPDATED: "calendarRsvpUpdated",
        CALENDAR_EVENT_RSVP_MANY_UPDATED: "calendarRsvpManyUpdated",
        CALENDAR_EVENT_RSVP_DELETED: "calendarRsvpDeleted",
        MESSAGE_CREATED: "messageCreated",
        MESSAGE_UPDATED: "messageUpdated",
        MESSAGE_DELETED: "messageDeleted",
        MEMBER_UPDATED: "memberUpdated",
        ROLES_UPATED: "rolesUpdated",
        MEMBER_JOINED: "memberJoined",
        MEMBER_REMOVED: "memberRemoved",
        MEMBER_BANNED: "memberBanned",
        MEMBER_UNBANNED: "memberUnbanned",
        WEBHOOK_CREATED: "webhookCreated",
        WEBHOOK_UPDATED: "webhookUpdated",
        LIST_ITEM_COMPLETED: "listItemCompleted",
        LIST_ITEM_UNCOMPLETED: "listItemUncompleted",
        LIST_ITEM_CREATED: "listItemCreated",
        LIST_ITEM_UPDATED: "listItemUpdated",
        LIST_ITEM_DELETED: "listItemDeleted",
        CHANNEL_CREATED: "channelCreated",
        CHANNEL_UPDATED: "channelUpdated",
        CHANNEL_DELETED: "channelDeleted",
        DOC_CREATED: "docCreated",
        DOC_UPDATED: "docUpdated",
        DOC_DELETED: "docDeleted",
        SERVER_CREATED: "serverCreated",
        BOT_SERVER_CREATED: "botServerCreated",
        MESSAGE_REACTION_CREATED: "messageReactionCreated",
        MESSAGE_REACTION_DELETED: "messageReactionDeleted",
        FORUM_TOPIC_CREATED: "forumTopicCreated",
        FORUM_TOPIC_DELETED: "forumTopicDeleted",
        FORUM_TOPIC_UPDATED: "forumTopicUpdated",
        FORUM_TOPIC_PINNED: "forumTopicPinned",
        FORUM_TOPIC_UNPINNED: "forumTopicUnpinned",
        FORUM_TOPIC_LOCKED: "forumTopicLocked",
        FORUM_TOPIC_UNLOCKED: "forumTopicUnlocked",
    },
};
exports.ROUTES = {
    AWS_CDN: "https://s3-us-west-2.amazonaws.com/www.guilded.gg/",
    BASE_DOMAIN: "www.guilded.gg",
    API_DOMAIN: "api.guilded.gg",
    CDN: "img.guildedcdn.com",
    IMAGE_CDN_DOMAIN: "img.guildedcdn.com",
    MEDIA_DOMAIN: "media.guilded.gg",
};
