// guilded.js/dist/managers/global/ForumManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalForumTopicManager = void 0;
const CacheableStructManager_1 = require("./CacheableStructManager");
require("./GlobalManager");
class GlobalForumTopicManager extends CacheableStructManager_1.CacheableStructManager {
    get shouldCacheForumTopic() {
        var _a;
        return ((_a = this.client.options.cache) === null || _a === void 0 ? void 0 : _a.cacheForumTopics) !== false;
    }
    /** Create a topic in a forum */
    createForumTopic(channelId, options) {
        return this.client.rest.router.createForumTopic(channelId, options).then((data) => data.forumTopic);
    }
    /** Get all topics in a forum */
    getForumTopics(channelId, options) {
        return this.client.rest.router.getForumTopics(channelId, options);
    }
    /** Get a topic in a forum */
    getForumTopic(channelId, forumThreadId) {
        return this.client.rest.router.getForumTopic(channelId, forumThreadId);
    }
    /** Update a topic in a forum */
    updateForumTopic(channelId, forumThreadId, options) {
        return this.client.rest.router.updateForumTopic(channelId, forumThreadId, options);
    }
    /** Delete a topic in a forum */
    deleteForumTopic(channelId, forumThreadId) {
        return this.client.rest.router.deleteForumTopic(channelId, forumThreadId);
    }
    /** Pin a topic in a forum */
    pinForumTopic(channelId, forumThreadId) {
        return this.client.rest.router.pinForumTopic(channelId, forumThreadId);
    }
    /** Unpin a topic in a forum */
    unpinForumTopic(channelId, forumThreadId) {
        return this.client.rest.router.unpinForumTopic(channelId, forumThreadId);
    }
    /** Lock a topic in a forum */
    lockForumTopic(channelId, forumThreadId) {
        return this.client.rest.router.lockForumTopic(channelId, forumThreadId);
    }
    /** Unlock a topic in a forum */
    unlockForumTopic(channelId, forumThreadId) {
        return this.client.rest.router.unlockForumTopic(channelId, forumThreadId);
    }
}
exports.GlobalForumTopicManager = GlobalForumTopicManager;
