// guilded.js/dist/gateway/handler/ReactionEventHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactionEventHandler = void 0;
const constants_1 = require("../../constants");
const structures_1 = require("../../structures");
const util_1 = require("../../util");
const GatewayEventHandler_1 = require("./GatewayEventHandler");
class ReactionEventHandler extends GatewayEventHandler_1.GatewayEventHandler {
    messageReactionCreated(data) {
        const newReaction = new structures_1.MessageReaction(this.client, {
            emote: data.d.reaction.emote,
            channelId: data.d.reaction.channelId,
            serverId: data.d.serverId,
            createdBy: data.d.reaction.createdBy,
            messageId: data.d.reaction.messageId,
        });
        if (this.client.reactions.shouldCacheReaction)
            this.client.reactions.cache.set((0, util_1.buildMessageReactionKey)(data.d.reaction.messageId, data.d.reaction.createdBy, data.d.reaction.emote.id), newReaction);
        return this.client.emit(constants_1.constants.clientEvents.MESSAGE_REACTION_CREATED, newReaction);
    }
    messageReactionDeleted(data) {
        if (this.client.reactions.shouldCacheReaction)
            this.client.reactions.cache.delete((0, util_1.buildMessageReactionKey)(data.d.reaction.messageId, data.d.reaction.createdBy, data.d.reaction.emote.id));
        return this.client.emit(constants_1.constants.clientEvents.MESSAGE_REACTION_DELETED, data.d);
    }
}
exports.ReactionEventHandler = ReactionEventHandler;
